datalad_next/iter_collections/gitworktree.py

```diff
@@ -0,0 +1,273 @@
+    :class:`GitWorktreeFileSystemItem` instances. In both cases, ``gitsha`` and
+    ``gittype`` properties are provided. Either of them being ``None``
+    indicates untracked work tree content.
+
+    .. note::
+      The ``gitsha`` is not equivalent to a SHA1 hash of a file's content,
+      but is the SHA-type blob identifier as reported and used by Git.
+
+    Parameters
+    ----------
+    path: Path
+      Path of a directory in a Git repository to report on. This directory
+      need not be the root directory of the repository, but must be part of
+      the repository's work tree.
+    untracked: {'all', 'whole-dir', 'no-empty'} or None, optional
+      If not ``None``, also reports on untracked work tree content.
+      ``all`` reports on any untracked file; ``whole-dir`` yields a single
+      report for a directory that is entirely untracked, and not individual
+      untracked files in it; ``no-empty-dir`` skips any reports on
+      untracked empty directories. Any untracked content is yielded as
+      a ``PurePosixPath``.
+    link_target: bool, optional
+      If ``True``, each file-type item includes a file-like object
+      to access the file's content. This file handle will be closed
+      automatically when the next item is yielded.
+    fp: bool, optional
+      If ``True``, each file-type item includes a file-like object
+      to access the file's content. This file handle will be closed
+      automatically when the next item is yielded.
+
+    Yields
+    ------
+    :class:`GitWorktreeItem` or `GitWorktreeFileSystemItem`
+    """
+    lsfiles_args = ['--stage', '--cached']
+    if untracked:
+        lsfiles_args.extend(lsfiles_untracked_args[untracked])
+
+    # helper to handle multi-stage reports by ls-files
+    pending_item = None
+
+    # we add a "fake" `None` record at the end to avoid a special
+    # case for submitting the last pending item after the loop.
+    # otherwise the context manager handling of the file pointer
+    # would lead to lots of code duplication
+    for line in chain(_git_ls_files(path, *lsfiles_args), [None]):
+        # a bit ugly, but we need to account for the `None` record
+        # that signals the final loop iteration
+        ipath, lsfiles_props = _lsfiles_line2props(line) \
+            if line is not None else (None, None)
+
+        # yield any pending item, if the current record is not an
+        # addendum of it
+        if ipath is None or (
+                pending_item is not None and pending_item[0] != ipath):
+            # report on a pending item, this is not a "higher-stage"
+            # report by ls-files
+            item = _get_item(path, link_target, fp, *pending_item)
+            if fp and item.type == FileSystemItemType.file:
+                with (Path(path) / item.name).open('rb') as fp:
+                    item.fp = fp
+                    yield item
+            else:
+                yield item
+            pending_item = None
```

Comment:
I think the assignment to `pending_item` is not used because `pending_item` is either set below in both branches of the if-clause, or the loop is left in line 181